### PR TITLE
fix: don't treat `\` as escape in standard strings, support E-strings, support vars after `->>` operator

### LINKED ISSUE #14693: String ending with backslash breaks PostgreSQL value replacements
Using PostgreSQL dialect replacements don't work if query contains a string with a backslash at the end.

## Issue Creation Checklist

- [x] I understand that my issue will be automatically closed if I don't fill in the requested information
- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Bug Description

When the query contains a fixed string that ends with a backslash all placeholders after the string aren't replaced anymore.

### Reproducible Example

```typescript
it("properly replaced placeholders", async () => {
  const result = await sequelize.query<{ backslash: string; replaced: string }>(
    "select '\\' as backslash, :test as replaced",
    {
      replacements: { test: "replaced" },
      type: QueryTypes.SELECT,
    }
  );
  expect(result).toEqual([{ baskslash: "\\", replaced: "replaced" }]);
});
```

### What do you expect to happen?

All placeholders get successfully replaced and the test above succeeds.

### What is actually happening?

Placeholders after the string which ends with a backslash aren't replaced resulting into a syntax error `syntax error at or near ":"`. The query that gets executed is this:
```sql
select '\' as backslash, :test as replaced
```

### Environment

- Sequelize version: 6.21.0
- Node.js version: 16.14.2
- If TypeScript related: TypeScript version: 4.7.4
- Database & Version: PostgreSQL 11.5
- Connector library & Version: pg 8.7.3

## Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I will need guidance.
- [ ] No, I don't have the time, but my company or I are [supporting Sequelize through donations on OpenCollective](https://opencollective.com/sequelize).
- [x] No, I don't have the time, and I understand that I will need to wait until someone from the community or maintainers is interested in resolving my issue.

---

*Indicate your interest in the resolution of this issue by adding the 👍 reaction. Comments such as "+1" will be removed.*

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Closes #14693

This PR fixes `injectReplacements` and `mapBindParameters` to support:
- "standard conforming strings", aka strings in which `\` cannot be used to escape `'`
- Postgres' E-prefixed strings (escape strings), which support `\`-based escaping, even if `standardConformingStrings` is true
- Parsing bind parameters and replacements after the JSONB `->` & `->>` operators

As a result, these two methods will now throw if the SQL includes unterminated string literals.

It was painful to test that one dialect threw (because it did not support `\` escaping), and another did not, so I introduced a new `expectPerDialect` utility based on `expectsql`

Here is such an example:

image

I also expanded `expectsql` to accept a combination of dialects separated by a space as the key, to reduce duplication (mainly for mysql/mariadb). e.g. the following is now possible:

```js
expectsql({
  default: 'SELECT ...',
  'mysql mariadb': 'SELECT ...',
})
```

### CODE REVIEW

**Comment by WikiRik** [comment]
Do we need a similar PR for v6?

**Comment by ephys** [comment]
Adding an extra fix

**Review by WikiRik — COMMENTED** [review]
Looks good! Few small questions

File: src/dialects/postgres/index.js [hunk 1]
```diff
@@ -72,6 +73,15 @@ export class PostgresDialect extends AbstractDialect {
     return createSpecifiedOrderedBindCollector();
   }
 
+  canBackslashEscape() {
+    // postgres can use \ to escape if one of these is true:
+    // - standard_conforming_strings is off
+    // - backslash_quote is on
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/postgres/index.js:79]: Turns out that statement is incorrect.

This results in a syntax error:

```sql
SET standard_conforming_strings = OFF;
SELECT '\';
```

but this does not:

```sql
SET backslash_quote = ON;
SELECT '\';
```

File: test/unit/utils/sql.test.ts [hunk 2]
```diff
@@ -656,3 +733,9 @@ describe('injectReplacements (positional replacements)', () => {
     });
   });
 });
+
+function getNonStandardConfirmingStringDialect() {
```
[the remark this answers is not in the document]
Comment (ephys) [reply test/unit/utils/sql.test.ts:768]: It's not really relevant elsewhere, no other test uses the `standardConformingStrings` option, and almost no code does either

File: src/utils/sql.ts [hunk 3]
```diff
@@ -99,6 +101,24 @@ function mapBindParametersAndReplacements(
 
     if (char === `'`) {
       isString = true;
+
+      // Fun fact, the following query is supported in almost all dialects,
+      //  SELECT E'test';
+      // but postgres interprets it as an E-prefixed string, while other dialects interpret it as
```
[the remark this answers is not in the document]
Comment (WikiRik) [reply src/utils/sql.ts:107]: Interesting

File: src/utils/sql.ts [hunk 4]
```diff
@@ -99,6 +101,24 @@ function mapBindParametersAndReplacements(
 
     if (char === `'`) {
       isString = true;
+
+      // Fun fact, the following query is supported in almost all dialects,
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/utils/sql.ts]: Done :)

File: src/dialects/postgres/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ephys) [reply src/dialects/postgres/index.js]: Removed the incorrect comment about backslash_quote

**Review by WikiRik — APPROVED** [review]

File: src/utils/sql.ts [hunk 5]
```diff
@@ -99,6 +101,25 @@ function mapBindParametersAndReplacements(
...
+      // which selects the type E and aliases it to 'test'.
+
+      stringIsBackslashEscapable
+        // all ''-style strings in this dialect can be backslash escaped
+        = dialect.canBackslashEscape()
+        // checking if this is a postgres-style E-prefixed string, which also supports backslash escaping
+        || (
+          dialect.supports.escapeStringConstants
+          // is this a E-prefixed string, such as `E'abc'` ?
+          && sqlString[i - 1] === 'E'
```
Comment (tomquist) [inline src/utils/sql.ts:118]: Shouldn't this also check for lowercase 'e'?

Comment (ephys) [reply src/utils/sql.ts:118]: Yes it should! Thanks for double checking :)

**Comment by WikiRik** [comment]
@sdepold you mentioned you wanted to backport this to v6 (together with #14733 ). Did you work on that already? Might be a nice PR to make for hacktoberfest

**Comment by sdepold** [comment]
Hey. Didn't have the time so far. Looking at this pr I got the impression that it was rather tricky but given that it's mostly a matter of replicating things, sure why not :)

**Pull request merged**